kuma/wiki/search.py

```diff
@@ -288,86 +287,39 @@ def reindex_all(cls, chunk_size=1000, index=None, percent=100):
-            'index.refresh_interval', settings.ES_DEFAULT_REFRESH_INTERVAL)
-        number_of_replicas = index_settings.get(
-            'number_of_replicas', settings.ES_DEFAULT_NUM_REPLICAS)
-
-        # Disable automatic refreshing.
-        temporary_settings = {
-            'index': {
-                'refresh_interval': '-1',
-                'number_of_replicas': '0',
-            }
-        }
+        total = len(indexable)
+        total_chunks = int(ceil(total / float(chunk_size)))
 
-        try:
-            es.indices.put_settings(temporary_settings, index=index_name)
-            start_time = time.time()
-
-            cls_start_time = time.time()
-            total = len(indexable)
-
-            if total == 0:
-                return
-
-            log.info('Reindex %s. %s to index...', cls.get_doc_type(), total)
-
-            i = 0
-            for chunk in chunked(indexable, chunk_size):
-                index_documents(chunk, index.pk)
-
-                i += len(chunk)
-                time_to_go = (total - i) * ((time.time() - start_time) / i)
-                per_chunk_size = ((time.time() - start_time) /
-                                  (i / float(chunk_size)))
-                log.info('%s/%s... (%s to go, %s per %s docs)', i, total,
-                         format_time(time_to_go),
-                         format_time(per_chunk_size),
-                         chunk_size)
-
-                # We call this every 1000 or so because we're
-                # essentially loading the whole db and if DEBUG=True,
-                # then Django saves every sql statement which causes
-                # our memory to go up up up. So we reset it and that
-                # makes things happier even in DEBUG environments.
-                reset_queries()
-
-            delta_time = time.time() - cls_start_time
-            log.info('Done! (%s, %s per %s docs)',
-                     format_time(delta_time),
-                     format_time(delta_time / (total / float(per_chunk_size))),
-                     chunk_size)
-
-        finally:
-            # Re-enable automatic refreshing
-            reset_settings = {
-                'index': {
-                    'refresh_interval': refresh_interval,
-                    'number_of_replicas': number_of_replicas,
-                }
-            }
-            es.indices.put_settings(reset_settings, index_name)
-            delta_time = time.time() - start_time
-            log.info('Done! (total time: %s)', format_time(delta_time))
+        pre_task = prepare_index.si(index.pk)
+        post_task = finalize_index.si(index.pk)
+
+        if not total:
+            # If there's no data we still create the index and finalize it.
+            chain(pre_task, post_task).apply_async()
+        else:
+            index_tasks = [index_documents.si(chunk, index.pk)
+                           for chunk in chunked(indexable, chunk_size)]
+            if settings.CELERY_ALWAYS_EAGER:
```

Comment:
I think it'd be neat to refactor that to a general helper function, e.g.:

```python
def flow(pre_task, main_tasks, post_task):
    if settings.CELERY_ALWAYS_EAGER:
        # Eager mode and chords don't get along. So we serialize
        # the tasks as a workaround.
        main_tasks.insert(0, pre_task)
        main_tasks.append(post_task)
        task = chain(*main_tasks)
    else:
        task = chain(pre_task, chord(header=main_tasks, body=post_task))
    return task
```

The use it like this:
```python
index_tasks = [index_documents.si(chunk, index.pk)
               for chunk in chunked(indexable, chunk_size)]

flow(pre_task, index_tasks, post_task).apply_async()
```